File: python/sdk/send_tx.py
from pwrpy.pwrwallet import Wallet
import time
from datetime import datetime, timedelta
from dotenv import load_dotenv
import os
load_dotenv()

# Setting up your wallet in the SDK
seed_phrase = os.getenv("SEED_PHRASE")
wallet = Wallet.new(seed_phrase)

def remove_guardian():
    # Remove your wallet guardian
    res = wallet.remove_guardian()

    # Error handling
    if res.success:
        print("Transaction Hash:", res.hash.hex())
    else:
        print("Error:", res.message)

def set_guardian():
    # Guardian address that will verify your transactions
    guardian = "0x34bfe9c609ca72d5a4661889033a221fa07ef61a"

    # Guardian validity period - 30 minutes
    current_time = datetime.now()
    future_time = current_time + timedelta(minutes=30) # 30 minutes from now
    expiry_date = int(time.mktime(future_time.timetuple()))

    # Set your wallet guardian
    res = wallet.set_guardian(expiry_date, guardian)

    # Error handling
    if res.success:
        print("Transaction Hash:", res.hash.hex())
    else:
        print("Error:", res.message)

def move_stake():
    from_validator = "FROM_VALIDATOR_ADDRESS"
    to_validator = "TO_VALIDATOR_ADDRESS"
    # Tokens amount - 1 PWR = 1e9 = 1000000000
    amount = 1000000000

    # Move stake token from validator to another
    res = wallet.move_stake(amount, from_validator, to_validator)

    # Error handling
    if res.success:
        print("Transaction Hash:", res.hash.hex())
    else:
        print("Error:", res.message)

def withdraw():
    # Validator address you delegated
    validator_address = "VALIDATOR_ADDRESS"
    # Tokens amount - 1 PWR = 1e9 = 1000000000
    amount = 1000000000

    # Withdraw the delegated pwr tokens
    res = wallet.withdraw(amount, validator_address)

    # Error handling
    if res.success:
        print("Transaction Hash:", res.hash.hex())
    else:
        print("Error:", res.message)

def delegate():
    # Validator address
    validator_address = "VALIDATOR_ADDRESS"
    # Tokens amount - 1 PWR = 1e9 = 1000000000
    amount = 1000000000

    # Delegate the validator
    res = wallet.delegate(validator_address, amount)

    # Error handling
    if res.success:
        print("Transaction Hash:", res.hash.hex())
    else:
        print("Error:", res.message)

def send_payable_data():
    # VM id used to send the transaction to
    vida_id = 123
    # Tokens amount - 1 PWR = 1e9 = 1000000000
    amount = 10
    # Buffer data to be included in the transaction
    data = "LFG!".encode()

    # Send the data at vmID 919 and pay 1e3
    res = wallet.send_payable_vida_data(vida_id, data, amount)

    # Error handling
    if res.success:
        print("Transaction Hash:", res.hash.hex())
    else:
        print("Error:", res.message)

def send_data():
    # VM id used to send the transaction to
    vida_id = 123
    # Buffer data to be included in the transaction
    data = "Hello World!".encode()

    # Send the data at vmID 123 to the chain
    res = wallet.send_vida_data(vida_id, data)

    # Error handling
    if res.success:
        print("Transaction Hash:", res.hash.hex())
    else:
        print("Error:", res.message)


def transfer():
    # Tokens recipient address
    recipient_address = "0x3B3B69093879E7B6F28366FA3C32762590FF547E"
    # Tokens amount - 1 PWR = 1e9 = 1000000000
    amount = 1000
    # Transfer pwr tokens from the wallet
    res = wallet.transfer_pwr(recipient_address, amount)

    # Error handling
    if res.success:
        print("Transaction Hash:", res.hash.hex())
    else:
        print("Error:", res.message)


if __name__ == "__main__":
    transfer()
    send_data()
    send_payable_data()
    # delegate()
    # withdraw()
    # move_stake()
    # set_guardian()
    # remove_guardian()

File: go/sdk/sendTx.go
package sdk

import (
	"fmt"
	"os"
	"time"

	"github.com/joho/godotenv"
	"github.com/pwrlabs/pwrgo/wallet"
)

func removeGuardian() {
	godotenv.Load()
	seedPhrase := os.Getenv("SEED_PHRASE")
	// Setting up your wallet in the SDK
	wallet, err := wallet.New(seedPhrase)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	feePerByte := wallet.GetRpc().GetFeePerByte()

	// Remove your wallet guardian
	tx := wallet.RemoveGuardian(feePerByte)

	// Error handling
	if tx.Success {
		fmt.Printf("Transaction Hash: %s\n", tx.Hash)
	} else {
		fmt.Println("Error:", tx.Error)
	}
}

func setGuardian() {
	godotenv.Load()
	seedPhrase := os.Getenv("SEED_PHRASE")
	// Setting up your wallet in the SDK
	wallet, err := wallet.New(seedPhrase)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Guardian address that will verify your transactions
	guardian := "0x34bfe9c609ca72d5a4661889033a221fa07ef61a"
	// Guardian validity period - 30 minutes
	futureDate := time.Now().Add(30 * time.Minute) // 30 minutes from now
	expiryDate := int(futureDate.Unix())           // Get the Unix timestamp in seconds

	feePerByte := wallet.GetRpc().GetFeePerByte()

	// Set your wallet guardian
	tx := wallet.SetGuardian(expiryDate, guardian, feePerByte)

	// Error handling
	if tx.Success {
		fmt.Printf("Transaction Hash: %s\n", tx.Hash)
	} else {
		fmt.Println("Error:", tx.Error)
	}
}

func moveStake() {
	godotenv.Load()
	seedPhrase := os.Getenv("SEED_PHRASE")
	// Setting up your wallet in the SDK
	wallet, err := wallet.New(seedPhrase)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	fromValidator := "FROM_VALIDATOR_ADDRESS"
	toValidator := "TO_VALIDATOR_ADDRESS"
	// Tokens amount - 1 PWR = 1e9 = 1000000000
	amount := int64(1e9)
	feePerByte := wallet.GetRpc().GetFeePerByte()

	// Move stake token from validator to another
	tx := wallet.MoveStake(amount, fromValidator, toValidator, feePerByte)

	// Error handling
	if tx.Success {
		fmt.Printf("Transaction Hash: %s\n", tx.Hash)
	} else {
		fmt.Println("Error:", tx.Error)
	}
}

func withdraw() {
	godotenv.Load()
	seedPhrase := os.Getenv("SEED_PHRASE")
	// Setting up your wallet in the SDK
	wallet, err := wallet.New(seedPhrase)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Validator address
	validator := "VALIDATOR_ADDRESS"
	// Tokens amount - 1 PWR = 1e9 = 1000000000
	amount := int(1e9)
	feePerByte := wallet.GetRpc().GetFeePerByte()

	// Withdraw the delegated pwr tokens
	tx := wallet.Withdraw(amount, validator, feePerByte)

	// Error handling
	if tx.Success {
		fmt.Printf("Transaction Hash: %s\n", tx.Hash)
	} else {
		fmt.Println("Error:", tx.Error)
	}
}

func delegate() {
	godotenv.Load()
	seedPhrase := os.Getenv("SEED_PHRASE")
	// Setting up your wallet in the SDK
	wallet, err := wallet.New(seedPhrase)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Validator address
	validator := "VALIDATOR_ADDRESS"
	// Tokens amount - 1 PWR = 1e9 = 1000000000
	amount := int(1e9)
	feePerByte := wallet.GetRpc().GetFeePerByte()

	// Delegate the validator
	tx := wallet.Delegate(validator, amount, feePerByte)

	// Error handling
	if tx.Success {
		fmt.Printf("Transaction Hash: %s\n", tx.Hash)
	} else {
		fmt.Println("Error:", tx.Error)
	}
}

func sendPayableData() {
	godotenv.Load()
	seedPhrase := os.Getenv("SEED_PHRASE")
	// Setting up your wallet in the SDK
	wallet, err := wallet.New(seedPhrase)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	// VM id used to send the transaction to
	vidaId := 919
	// Tokens amount - 1 PWR = 1e9 = 1000000000
	amount := int64(10)
	// Buffer data to be included in the transaction
	data := []byte("Hello world")
	feePerByte := wallet.GetRpc().GetFeePerByte()

	// Send the data at vmID 919 and pay 1e3
	tx := wallet.SendPayableVidaData(vidaId, data, amount, feePerByte)

	// Error handling
	if tx.Success {
		fmt.Printf("Transaction Hash: %s\n", tx.Hash)
	} else {
		fmt.Println("Error:", tx.Error)
	}
}

func sendData() {
	godotenv.Load()
	seedPhrase := os.Getenv("SEED_PHRASE")
	// Setting up your wallet in the SDK
	wallet, err := wallet.New(seedPhrase)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	// VM id used to send the transaction to
	vidaId := 123
	// Buffer data to be included in the transaction
	data := []byte("Hello world")
	feePerByte := wallet.GetRpc().GetFeePerByte()

	// Send the data at vmID 123 to the chain
	tx := wallet.SendVidaData(vidaId, data, feePerByte)

	// Error handling
	if tx.Success {
		fmt.Printf("Transaction Hash: %s\n", tx.Hash)
	} else {
		fmt.Println("Error:", tx.Error)
	}
}

func transfer() {
	godotenv.Load()
	seedPhrase := os.Getenv("SEED_PHRASE")
	// Setting up your wallet in the SDK
	wallet, err := wallet.New(seedPhrase)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Tokens recipient address
	recipientAddress := "0x3B3b69093879e7B6F28366Fa3c32762590Ff547e"
	// Tokens amount - 1 PWR = 1e9 = 1000000000
	amount := int(1e3)
	feePerByte := wallet.GetRpc().GetFeePerByte()

	// Transfer pwr tokens from the wallet
	tx := wallet.TransferPWR(recipientAddress, amount, feePerByte)

	// Error handling
	if tx.Success {
		fmt.Printf("Transaction Hash: %s\n", tx.Hash)
	} else {
		fmt.Println("Error:", tx.Error)
	}
}

func SendTx() {
	// transfer()
	// sendData()
	// sendPayableData()
	// delegate()
	// withdraw()
	// moveStake()
	// setGuardian()
	removeGuardian()
}
